Streams/FileSliceReader.cpp

```diff
@@ -59,23 +74,23 @@ void FileSliceReader::SeekRelative(int64_t offset)
 	fileStreamReader.SeekRelative(offset);
 }
 
-std::unique_ptr<SeekableStreamReader> FileSliceReader::Slice(uint64_t sliceLength) 
+FileSliceReader FileSliceReader::Slice(uint64_t sliceLength) 
 {
-	auto slice = Slice(startingOffset + Position(), sliceLength);
+	FileSliceReader slice = Slice(startingOffset + Position(), sliceLength);
```

Comment:
I'm trying to decide when to use auto.

http://www.acodersjourney.com/2016/02/c-11-auto/

Reply:
Lol, nice meme.

I found the comments at the end of the article particularly interesting. I think I agree with some of those comments more so than the negative points in the article. In particular, not needing to know or care what a type is in many cases, and the type is much less important than a good descriptive name that tells you what the value represents. Knowing what the value represents provides a good clue as to how a value should be used.

Compare the following:
```
auto str = SomeFunction();  // Name indicates type, but it otherwise vague
std::string a = SomeFunction();  // Type known, but name is vague
auto filename = SomeFunction();  // Name implies usage (which may suggest possible types)
```
I'd say the last is the most useful, and would be useful if the underlying type was a `std::string`, or if it was a `std::filesystem::path`. Further, chances are, surrounding code wouldn't care which of the two it was, leaving open the possibility of changing the function signature with little impact on the rest of the code.

Worth considering, is languages that don't have strict typing. In particular, scripting languages like Ruby or Python. Python is known for duck typing (if it quacks like a duck, it's a duck). You rely on good descriptive names, because there is no fixed type. C++ templates basically work with duck typing. Provided a type supports all the needed operations performed on it, the template can be instantiated for that type. What matters is how it's used, more so than the exact type.